Add type hints from typeshed

phershbe opened the pull request:
Issue: https://github.com/jazzband/prettytable/issues/203

I didn't do the third part yet...
**Add mypy via pre-commit and make mypy pass (can be follow-up PR if there's much to do)**
...as I'm newer to contributing and didn't want to get overwhelmed. I can open a new issue so that it doesn't get forgotten and work on that next.

Used merge_pyi to integrate the type annotations and added appropriate args to .pre-commit-config.yaml in order for pre-commit to add the import to Python files so that newer typing syntax can be used on older Python 3.7 (as described in the issue).

Note: Did not use Black to format as it was deleting spaces in the docstrings which presumably are supposed to be there; can use Black and then add the spaces again after if that's better.

phershbe committed:
Integrated with typeshed stub file

phershbe committed:
Integrated with typeshed stub file

phershbe committed:
Added args to isort

As per https://github.com/jazzband/prettytable/issues/203

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

hugovk commented on src/prettytable/prettytable.py:
```diff
@@ -41,8 +43,10 @@
+from typing import Any
 
 import wcwidth
+from _typeshed import Incomplete
```
We don't have access to this `_typeshed` module so need something else.

It's an alias for `typing.Any`:

https://github.com/python/typeshed/blob/f9ed090cfc054fa7393ad1affb30e76f238871f8/stdlib/_typeshed/__init__.pyi#L32-L37

Generally it's best to avoid `Any` and use something more specific, but for a first iteration let's use `Any` (already imported above).

```suggestion
```

hugovk commented on src/prettytable/colortable.py:
```diff
@@ -1,3 +1,10 @@
+from typing import ClassVar
+
+from _typeshed import Incomplete
+
```
These aren't used in this file, let's remove them:

```suggestion
```

hugovk commented on src/prettytable/prettytable.py:
```diff
@@ -558,11 +562,11 @@ def _validate_attributes(self, name, val):
     ##############################
     @property
-    def rows(self):
+    def rows(self) -> list[Incomplete]:
```
```suggestion
    def rows(self) -> list[Any]:
```

hugovk reviewed (commented):
Thank you! 

> I didn't do the third part yet... **Add mypy via pre-commit and make mypy pass (can be follow-up PR if there's much to do)** ...as I'm newer to contributing and didn't want to get overwhelmed. I can open a new issue so that it doesn't get forgotten and work on that next.

Yep, let's keep the scope narrower for this first PR. 👍 

Here's some review comments.

phershbe committed:
Update src/prettytable/prettytable.py

Remove _typeshed reference as we don't have access to it and we can use typing.Any instead.

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

phershbe committed:
Update src/prettytable/colortable.py

Removed Callable, ClassVar, and Incomplete since they aren't used in the file.

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

phershbe committed:
Update src/prettytable/prettytable.py

Changed Incomplete to Any since we don't have access to the _typeshed module.

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

phershbe committed:
Removed import of List

Removed import of "List" from "typing" since we now have access to "list" from the "from __future__ import annotations" line

hugovk reviewed (approved):
Thank you for your contribution!
